### ISSUE TITLE
mel_to_stft and _nnls_lbfgs_block

### ISSUE THREAD

**Issue by spyroot** [librosa/librosa#1494]
Hi Folks,

I guess it is more of a question. I tried to decipher SVD and _nnls_lbfgs_block.  I ported some code to GPU because 
I execute some librosa logic.  Does that block the account batch dimension?   in essence, if an input mel [64, 531, n]

i.e this one a tricky :)

```
    for bl_s in range(0, x.shape[-1], n_columns):
        bl_t = min(bl_s + n_columns, B.shape[-1])
        x[..., bl_s:bl_t] = _nnls_lbfgs_block(
            A, B[..., bl_s:bl_t], x_init=x_init[..., bl_s:bl_t], **kwargs
        )
```

Thank you

**Comment by bmcfee** [comment]
This loop is iterating over blocks of frames, processing `n_columns` at a time.  For each block, it compute non-negative least squares solution to `|Ax - B|²`.  This blocking is mainly done to keep memory usage from blowing up on us.  The `...` business is to support arbitrarily many leading axes, eg for mini-batches, multichannel or microphone array configurations, or any combination thereof.

If by SVD, you're referring to the pseudo-inverse a few lines up:
https://github.com/librosa/librosa/blob/7ac022a8496126e95b46a7d56ad880a328359bda/librosa/util/_nnls.py#L148-L150
this is done for initialization.  A pseudo-inverse won't respect non-negativity (which we need for a magnitude spectrogram, and nnls generally) but pinv followed by thresholding gets us a good initial point for the subsequent quasi-newton optimization (LBFGS).

**Comment by spyroot** [comment]
Yeah I manage to write half of bound L-BFGS slowly slowly porting : ) Initial code Fortran :))))

**Comment by bmcfee** [comment]
> Note I check how Torch does that ( they use SGD and it off) Libros implementation is much much better so much respect from my side.

LBFGS is going to be much, much more efficient for this sort of problem (strongly convex objective + convex constraints) than projected (thresholded) SGD.  Porting LBFGS is going to be a bit of a lift though.  I expect you could probably get close with minimal effort by swapping out SGD for adagrad.  You'll need to step up the number of iterations considerably (compared to lbfgs) but it might end up about equal if gpu parallelism can pay off.  You might also need to increase the number of frames you process at once to see any gains here, but this will be device specific.
